model/src/main/kotlin/DependencyNavigator.kt

```diff
@@ -95,9 +97,69 @@ interface DependencyNavigator {
         matcher: DependencyMatcher = MATCH_ALL
     ): Set<Identifier>
 
+    /**
+     * Return a map with the shortest paths to each dependency in all scopes of the given [project]. The path to a
+     * dependency is defined by the nodes of the dependency tree that need to be passed to get to the dependency. For
+     * direct dependencies the shortest path is empty. The resulting map has scope names as keys; the values are maps
+     * with the shorted paths to all the dependencies contained in that scope.
+     */
+    fun getShortestPaths(project: Project): Map<String, Map<Identifier, List<Identifier>>> =
+        getShortestPathForProject(scopeDependencies(project)) { directDependencies(project, it) }
+
     /**
      * Return the set of [Identifier]s that refer to sub-projects of the given [project].
      */
     fun collectSubProjects(project: Project): SortedSet<Identifier> =
         scopeDependencies(project, matcher = MATCH_SUB_PROJECTS).collectDependencies()
+
+    /**
+     * Determine the map of shortest paths for all the dependencies of a project, given its map of [scopeDependencies]
+     * and a [scopeResolver] function that retrieves the direct dependency nodes of a scope.
+     */
+    private fun getShortestPathForProject(
+        scopeDependencies: Map<String, Set<Identifier>>,
+        scopeResolver: (String) -> Sequence<DependencyNode>
+    ): Map<String, Map<Identifier, List<Identifier>>> =
```

Comment:
Could this be a memory issue for projects with many scopes and large dependency trees (e.g. Android projects), compared to having only the map for one scope at a time? I'm asking because I didn't check if there was a place where they were all in memory at the same time also with the old code.

Reply:
This is definitive something we need to have an eye on. I think the old code mostly used nested loops to iterate over all the scopes of a project and then over the dependencies of a scope; so the access here was sequential.

What I have done for now, is to introduce a new `dependenciesForScope()` function in `DependencyNavigator`, which only fetches the dependencies of a single scope. (There is a default implementation of `scopeDependencies()`, which constructs a map of all scopes.) `getShortestPathsForProject()` now uses this new function and therefore should behave analogously to the previous implementation. We can think about changing the other callers of `scopeDependencies()` in a similar way if necessary.